### PR TITLE
Device Configs Backup Module - 3 Bug Fixes

### PR DESCRIPTION
**Pull request by rukapse**
## Description
Please include a summary of the changes and the related issue. Also, include relevant motivation and context.

## Type of Change
- [ yes] Bug fix
- [yes ] New feature
- [no ] Breaking change
- [yes ] Documentation update

## Checklist
- [yes ] My code follows the style guidelines of this project
- [ yes] I have performed a self-review of my own code
- [yes ] I have commented my code, particularly in hard-to-understand areas
- [yes ] I have made corresponding changes to the documentation
- [yes ] My changes generate no new warnings
- [ yes] I have added tests that prove my fix is effective or that my feature works
- [pending ] New and existing unit tests pass locally with my changes
- [ yes] Any dependent changes have been merged and published in downstream modules
- [ yes] All the sanity checks have been completed and the sanity test cases have been executed

## Ansible Best Practices
- [yes ] Tasks are idempotent (can be run multiple times without changing state)
- [yes] Variables and secrets are handled securely (e.g., using `ansible-vault` or environment variables)
- [yes] Playbooks are modular and reusable
- [yes] Handlers are used for actions that need to run on change

## Documentation
- [yes ] All options and parameters are documented clearly.
- [yes ] Examples are provided and tested.
- [yes ] Notes and limitations are clearly stated.

## Notes to Reviewers

Changes Include:

1. device_configs_backup_workflow_manager.py
    **Bug Fix:** 
      Backup caps at 500 devices due to `get_device_list` api limit
    **Enhancements:** 
      Add input list for device_backup
      Configs export of devices using site name
2. Includes Integration Tests for the Device Configs Backup Integration Module

### CODE REVIEW

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 1]
```diff
@@ -483,320 +874,485 @@ def password_generator(self):
...
+
+        while True:
+            # Retrieve the task status using the task ID
+            response = self.get_task_status(task_id, task_name)
+
+            # Check if response returned
+            if not response:
+                self.msg = "Error retrieving Task status for the task_name {0} task_id {1}".format(task_name, task_id)
+                self.update_result("failed", False, self.msg, "ERROR")
+                result = False
```
Comment (madhansansel) [inline plugins/modules/device_configs_backup_workflow_manager.py:1119]: We don't need result, just return self from here..

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 2]
```diff
@@ -483,320 +874,485 @@ def password_generator(self):
...
+                        .format(task_name, self.want.get("export_device_configurations_params"))
+                    )
+                    self.update_result("failed", False, self.msg, "ERROR")
+                    result = False
+                    break
+
+            # Check if task completed successfully and exit the loop
+            if not response.get("isError") and response.get("progress") == "Device configuration Successfully exported as password protected ZIP.":
+                self.log("{0} Task completed successfully. Exiting the loop.".format(task_name), "INFO")
+                result = True
```
Comment (madhansansel) [inline plugins/modules/device_configs_backup_workflow_manager.py:1152]: no need of setting result, just break.

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 3]
```diff
@@ -483,320 +874,485 @@ def password_generator(self):
...
+                break
+
+            self.log("The progress status is {0}, continue to check the status after 3 seconds. Putting into sleep for 3 seconds.".format(
+                response.get("progress")), "INFO")
+            time.sleep(3)
+
+        # Perform additional tasks after breaking the loop
+        if result:
```
Comment (madhansansel) [inline plugins/modules/device_configs_backup_workflow_manager.py:1160]: Checking this flag is not needed.

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 4]
```diff
@@ -483,320 +874,485 @@ def password_generator(self):
...
+            mgmt_ip_to_instance_id_map = self.want.get("mgmt_ip_to_instance_id_map")
+            additionalStatusURL = response.get("additionalStatusURL")
```
Comment (madhansansel) [inline plugins/modules/device_configs_backup_workflow_manager.py:1162]: additional_status_url =

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 5]
```diff
@@ -192,287 +392,478 @@ class Device_configs_backup(DnacBase):
...
+            response = self.dnac._exec(
+                family="sites",
+                function="get_membership",
+                op_modifies=True,
+                params=site_params,
+            )
+            self.log("Response received post 'get_membership' API Call: {0} ".format(str(response)), "DEBUG")
+
+            # Process the response if available
+            if response:
```
[the remark this answers is not in the document]
Comment (rukapse) [reply plugins/modules/device_configs_backup_workflow_manager.py:582]: Already logging!

            else:
                # If unable to retrieve device information, log an error message
                self.log("No response received from API call 'get_membership' to get membership information for site. {0}".format(site_name), "ERROR")

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 6]
```diff
@@ -483,320 +874,485 @@ def password_generator(self):
...
-        return task_id
+        # Log the error if an exception occurs during the API call
+        except Exception as e:
+            self.msg = "Error occurred while retrieving 'get_task_by_id' for Task {0} with Task id {1}. Error: {2}".format(
+                task_name, task_id, str(e))
+            self.update_result("failed", False, self.msg, "ERROR")
+            self.check_return_status()
 
-    def get_task_status(self, task_id=None):
+    def exit_while_loop(self, start_time, task_id, task_name, response):
```
[the remark this answers is not in the document]
Comment (rukapse) [reply plugins/modules/device_configs_backup_workflow_manager.py:994]: I haven't pulled the latest changes and dnac.py has been changed, I will get conflicts if I change dnac.py right now. 
Will implement this change in the next PR.

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 7]
```diff
@@ -483,320 +874,485 @@ def password_generator(self):
...
-            self.log("File couldn't be downloaded: {0}".format(e), "INFO")
-            return False
+            self.msg = "The Backup Config file with File ID: {0} could not be downloaded due to the following error: {1}".format(file_id, e)
+            self.update_result("failed", False, self.msg, "ERROR")
+            self.check_return_status()
 
-        if isinstance(response, dict) and response.get("errorCode"):
-            self.log(response.get("message"), "CRITICAL")
-            self.module.fail_json(msg=response.get("message"))
+    def unzip_data(self, file_id, file_data):
```
[the remark this answers is not in the document]
Comment (rukapse) [reply plugins/modules/device_configs_backup_workflow_manager.py:1058]: I haven't pulled the latest changes and dnac.py has been changed, I will get conflicts if I change dnac.py right now. 
Will implement this change in the next PR.

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 8]
```diff
@@ -483,320 +874,485 @@ def password_generator(self):
...
         except Exception as e:
-            self.log("Error in unzipping: {0}".format(e), "CRITICAL")
-            return False
+            self.msg = "Error in unzipping Backup Config file with file ID: {0}. Error: {1}".format(file_id, e)
+            self.update_result("failed", False, self.msg, "ERROR")
+            self.check_return_status()
 
-        self.log("Unzipping complete", "INFO")
-        return True
+    def get_export_device_config_task_status(self, task_id):
```
[the remark this answers is not in the document]
Comment (rukapse) [reply plugins/modules/device_configs_backup_workflow_manager.py:1095]: I haven't pulled the latest changes and dnac.py has been changed, I will get conflicts if I change dnac.py right now. 
Will implement this change in the next PR.

File: plugins/modules/device_configs_backup_workflow_manager.py [hunk 9]
```diff
@@ -192,287 +392,478 @@ class Device_configs_backup(DnacBase):
...
+            get_device_list_params = self.get_device_list_params(config)
+            if get_device_list_params:
+                self.log("Attempting to get Device Id(s) of all device(s) using parameters(excluding site_list): {0}".format(
+                    get_device_list_params), "DEBUG")
+                params_mgmt_ip_to_instance_id_map = self.get_device_ids_by_params(get_device_list_params)
+                mgmt_ip_to_instance_id_map.update(params_mgmt_ip_to_instance_id_map)
+                self.log("Retrieved following Device Id(s) of device(s): {0} from the provided parameters(excluding site_list).".format(
+                    mgmt_ip_to_instance_id_map), "DEBUG")
+        else:
+            # If no site_list is provided, use other parameters to get device IDs
```
[the remark this answers is not in the document]
Comment (rukapse) [reply plugins/modules/device_configs_backup_workflow_manager.py:789]: Unfortunately, this cannot be changed cause the behaviour of the function is


if site_list: 
      gets all devices in sites specified
      
      if other params passed:
      gets devices based on specified params
      
else:
      gets devices based params except for the site_list
      Now in this we don't check if  other_params provided or not, cause if params not provided we get all the devices in the dnac.

**Review by madhansansel — APPROVED** [review]

**Review by madhansansel — APPROVED** [review]

**Pull request merged**